src/libcore/iter.rs

```diff
@@ -8,51 +8,278 @@
+//! and what the compiler converts it into:
+//!
+//! [`IntoIterator`]: trait.IntoIterator.html
+//! [`into_iter()`]: trait.IntoIterator.html#tymethod.into_iter
+//!
+//! ```
+//! let values = vec![1, 2, 3, 4, 5];
+//!
+//! for x in values {
+//!     println!("{}", x);
+//! }
+//! ```
+//!
+//! Rust de-sugars this into:
+//!
+//! ```
+//! let values = vec![1, 2, 3, 4, 5];
+//! {
+//!     let result = match values.into_iter() {
+//!         mut iter => loop {
+//!             match iter.next() {
+//!                 Some(x) => { println!("{}", x); },
+//!                 None => break,
+//!             }
+//!         },
+//!     };
+//!     result
+//! }
+//! ```
+//!
+//! First, we call `into_iter()` on the value. Then, we match on the iterator that returns, calling
+//! [`next()`] over and over until we see a `None`. At that point, we `break` out of the loop, and
+//! we're done iterating.
+//!
+//! There's one more subtle bit here: the standard library contains an interesting implementation
+//! of [`IntoIterator`]:
+//!
+//! ```ignore
+//! impl<I> IntoIterator for I where I: Iterator
+//! ```
+//!
+//! In other words, all [`Iterator`]s implement [`IntoIterator`], by just returning themselves.
+//! This means two things:
+//!
+//! 1. If you're writing an [`Iterator`], you can use it with a `for` loop.
+//! 2. If you're creating a collection, implementing [`IntoIterator`] for it will allow your
+//!    collection to be used with the `for` loop.
+//!
+//! # Adapters
+//!
+//! Functions which take an [`Iterator`] and return another [`Iterator`] are often called 'iterator
+//! adapters', as they're a form of the 'adapter pattern'.
+//!
+//! Common iterator adapters include [`map()`], [`take()`], and [`collect()`]. For more, see their
+//! documentation.
+//!
+//! [`map()`]: trait.Iterator.html#method.map
+//! [`take()`]: trait.Iterator.html#method.take
+//! [`collect()`]: trait.Iterator.html#method.collect
+//!
+//! # Laziness
+//!
+//! Iterators (and iterator [adapters](#adapters)) are *lazy*. This means that just creating an
+//! iterator doesn't _do_ a whole lot. Nothing really happens until you call [`next()`]. This is
+//! sometimes a source of confusion when creating an iterator solely for its side effects. For
+//! example, the [`map()`] method calls a closure on each element it iterates over:
+//!
+//! ```
+//! let v = vec![1, 2, 3, 4, 5];
+//! v.iter().map(|x| println!("{}", x));
+//! ```
+//!
+//! This will not print any values, as we only created an iterator, rather than using it. The
+//! compiler will warn us about this kind of behavior:
+//!
+//! ```text
+//! warning: unused result which must be used: iterator adaptors are lazy and do nothing unless
+//! consumed
+//! ```
+//!
+//! The idiomatic way to write a [`map()`] for its side effects is to use a `for` loop instead:
+//!
+//! ```
+//! let v = vec![1, 2, 3, 4, 5];
+//!
+//! for x in &v {
```

Comment:
I think it was just because that's usually what you want. And it's the same as map, which does by reference too. I think it's a bit okay to be not quite as linear in API docs.